File: tests/transport_reference_interop.py
#!/usr/bin/env python3
"""Cross-check PHP RC4/LZ77/KBin wire format against reference protocol.py."""

from __future__ import annotations

import sys
from pathlib import Path
from xml.etree import ElementTree as ET

INFO = "1-01234567-89ab"
FIXTURE = (
    '<?xml version="1.0" encoding="UTF-8"?>'
    '<call model="VFG:J:A:A:2025122300">'
    '<eventlog method="write">'
    '<gamesession __type="s64">9223372036854775807</gamesession>'
    '<message __type="str">麻雀ファイトガール</message>'
    '<globalip __type="ip4">127.0.0.1</globalip>'
    '</eventlog></call>'
)


def assert_xml(text: str) -> None:
    root = ET.fromstring(text)
    assert root.tag == "call"
    assert root.attrib.get("model") == "VFG:J:A:A:2025122300"
    eventlog = root.find("eventlog")
    assert eventlog is not None and eventlog.attrib.get("method") == "write"
    assert eventlog.findtext("gamesession") == "9223372036854775807"
    assert eventlog.findtext("message") == "麻雀ファイトガール"
    assert eventlog.findtext("globalip") == "127.0.0.1"


def main() -> None:
    if len(sys.argv) != 4:
        raise SystemExit(
            "usage: transport_reference_interop.py REFERENCE_DIR PHP_WIRE PYTHON_WIRE"
        )
    reference = Path(sys.argv[1]).resolve()
    sys.path.insert(0, str(reference))
    import protocol  # type: ignore  # reference repository module

    php_wire = Path(sys.argv[2]).read_bytes()
    text, meta = protocol.decode_eamuse_body(php_wire, INFO, "lz77")
    assert meta.used_kbin
    assert str(meta.kbin_encoding).upper().replace("_", "-") in {"UTF-8", "UTF8"}
    assert bool(meta.kbin_compressed) is True
    assert_xml(text)

    python_wire = protocol.encode_eamuse_body(
        FIXTURE,
        INFO,
        "lz77",
        protocol.EamuseDecodeMeta(
            used_kbin=True,
            kbin_encoding="UTF-8",
            kbin_compressed=True,
        ),
    )
    Path(sys.argv[3]).write_bytes(python_wire)
    print("PHP <-> Python reference RC4/LZ77/KBin transport OK")


if __name__ == "__main__":
    main()

File: tests/transport_reference_interop.php
<?php

declare(strict_types=1);

spl_autoload_register(static function(string $class):void{
    if(!str_starts_with($class,'Mfg\\'))return;
    $path=dirname(__DIR__).'/src/'.str_replace('\\','/',substr($class,4)).'.php';
    if(is_file($path))require $path;
});

use Mfg\Protocol\EamuseProtocol;
use Mfg\Protocol\KBinXml;

const TR_INFO='1-01234567-89ab';

function tr_ok(bool $v,string $m):void{if(!$v)throw new RuntimeException($m);}
function tr_xml():string{return '<?xml version="1.0" encoding="UTF-8"?><call model="VFG:J:A:A:2025122300"><eventlog method="write"><gamesession __type="s64">9223372036854775807</gamesession><message __type="str">麻雀ファイトガール</message><globalip __type="ip4">127.0.0.1</globalip></eventlog></call>';}
function tr_assert(string $xml):void{
    $d=new DOMDocument();tr_ok(@$d->loadXML($xml,LIBXML_NONET),'decoded XML invalid');
    $r=$d->documentElement;tr_ok($r instanceof DOMElement&&$r->tagName==='call','root');
    tr_ok($r->getAttribute('model')==='VFG:J:A:A:2025122300','model');
    $e=$d->getElementsByTagName('eventlog')->item(0);tr_ok($e instanceof DOMElement&&$e->getAttribute('method')==='write','eventlog');
    tr_ok($d->getElementsByTagName('gamesession')->item(0)?->textContent==='9223372036854775807','s64');
    tr_ok($d->getElementsByTagName('message')->item(0)?->textContent==='麻雀ファイトガール','UTF-8 string');
    tr_ok($d->getElementsByTagName('globalip')->item(0)?->textContent==='127.0.0.1','ip4');
}

$mode=$argv[1]??'';$path=$argv[2]??'';tr_ok($path!=='','usage: php transport_reference_interop.php emit|check FILE');
if($mode==='emit'){
    $kbin=KBinXml::encode(tr_xml(),'UTF-8',true);
    $wire=EamuseProtocol::encodeTransport($kbin,TR_INFO,'lz77');
    tr_ok(file_put_contents($path,$wire)!==false,'cannot write PHP transport');
    $round=EamuseProtocol::decodeTransport($wire,TR_INFO,'lz77');
    $meta=KBinXml::decode($round);tr_assert($meta['xml']);
    echo "PHP RC4/LZ77/KBin transport emitted\n";exit(0);
}
if($mode==='check'){
    $wire=file_get_contents($path);tr_ok($wire!==false,'cannot read Python transport');
    $kbin=EamuseProtocol::decodeTransport($wire,TR_INFO,'lz77');
    tr_ok(KBinXml::isBinary($kbin),'Python transport did not decode to KBin');
    $meta=KBinXml::decode($kbin);tr_ok($meta['encoding']==='UTF-8'&&$meta['compressed']===true,'Python KBin metadata');
    tr_assert($meta['xml']);
    echo "Python reference RC4/LZ77/KBin transport decoded by PHP\n";exit(0);
}
throw new RuntimeException('unknown mode '.$mode);
